src/json-patch.ts

```diff
@@ -289,59 +392,133 @@ module jsonpatch {
             }
-            results[p - 1] = arrOps[patch.op].call(patch, obj, key, tree); // Apply patch
-            break;
+            returnValue.result = arrOps[operation.op].call(operation, obj, key, document); // Apply patch
+            returnValue.newDocument = document;
+            return returnValue;
           }
         }
         else {
-          if (key && key.indexOf('~') != -1)
-            key = key.replace(/~1/g, '/').replace(/~0/g, '~'); // escape chars
+          if (key && key.indexOf('~') != -1) {
+            key = unEscapePath(key);
+          }
           if (t >= len) {
-            results[p - 1] = objOps[patch.op].call(patch, obj, key, tree); // Apply patch
-            break;
+            returnValue.result = objOps[operation.op].call(operation, obj, key, document); // Apply patch
+            returnValue.newDocument = document;
+            return returnValue;
           }
         }
         obj = obj[key];
       }
     }
+  }
+
+  /**
+   * Apply a JSON Patch on a JSON document.
+   * Returns an array of results of operations.
+   * Each element can either be a boolean (if op == 'test') or
+   * the removed object (operations that remove things)
+   * or just be undefined
+   */
+  export function applyPatch(document: any, patch: Operation[], validate?: boolean): OperationResult[] {
+    const results: OperationResult[] = new Array(patch.length);
+
+    for (let i = 0, length = patch.length; i < length; i++) {
+      results[i] = applyOperation.call(this, document, patch[i], validate, true);
+      document = results[i].newDocument; // in case root was replaced
+    }
     return results;
   }
 
+  /**
+   * Apply a JSON Patch on a JSON document.
+   * Returns an array of results of operations.
+   * Each element can either be a boolean (if op == 'test') or
+   * the removed object (operations that remove things)
+   * or just be undefined
+   * @deprecated
+   */
+  export function apply(document: any, patch: Operation[], validate?: boolean): any[] {
+    console.warn('jsonpatch.apply is deprecated, please use `applyPatch` for applying patch sequences, or `applyOperation` to apply individual operations.');
+    const results = new Array(patch.length);
+
+    /* this code might be overkill, but will be removed soon, it is to prevent the breaking change of root operations */
+    for (let i = 0, length = patch.length; i < length; i++) {
+
+      if (patch[i].path == "" && patch[i].op != "remove" && patch[i].op != "test") {
+        let value;
+
+        if (patch[i].op == "replace" || patch[i].op == "move") {
+          results[i] = deepClone(document);
+        }
+
+        if (patch[i].op == "copy" || patch[i].op == "move") {
+          value = jsonpatch.getValueByPointer(document, (<any>patch[i]).from);
+        }
+        if (patch[i].op == "replace" || patch[i].op == "add") {
+          value = (<any>patch[i]).value;
```

Comment:
Yeah, I though it might have been fixed. I would cast to `ReplaceOperation  | AddOperation` then